lib/linked_list.rb

```diff
@@ -19,38 +19,88 @@ def initialize
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
     def add_first(value)
-      raise NotImplementedError
+      @head = Node.new(value, @head)
     end
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
     def search(value)
-      raise NotImplementedError
+      current = @head 
+
+      until current.nil?
+        return true if current.data == value 
+        current = current.next
+      end 
+
+      return false
     end
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max
-      raise NotImplementedError
+      return nil if @head.nil?
+      max = 0
+      current = @head
+
+      until current.nil? 
+        if current.data > max
+          max = current.data
+        end 
+        current = current.next
+      end 
+      
+      return max
     end
 
     # method to return the min value in the linked list
     # returns the data value and not the node
     def find_min
-      raise NotImplementedError
+      return nil if @head.nil?
+      min = nil 
+      current = @head
+
+      until current.nil? 
+        if min == nil 
+          min = current.data 
+        elsif current.data < min 
+          min = current.data
+        end 
+        current = current.next
+      end
+
+      return min
     end
 
 
     # method that returns the length of the singly linked list
     def length
-      raise NotImplementedError
+      count = 0 
+      return count if @head.nil?
+
+      current = @head 
+      until current.nil?
+        current = current.next
+        count += 1 
+      end 
+
+      return count 
     end
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns nil if there are fewer nodes in the linked list than the index value
     def get_at_index(index)
-      raise NotImplementedError
+      return nil if @head.nil?
+
+      current = @head 
+      count = 0 
+
+      until count == index 
+        current = current.next 

```

Comment:
What about if the length is less than `index`?

```suggestion
        return nil if current.nil?
        current = current.next 
```